Improve qvm-ls options to avoid parsing

alimirjamali opened the pull request:
fixes: https://github.com/QubesOS/qubes-issues/issues/8614

Filtering, Formatting & Sorting ArgumentParser Groups
Sorting options based on output columns
A new pref(erence) format
Dedicated filter options for class, label, template, netvm preferences
Dedicated filter options for internal,servicevm & updates-available
Generic filtering options for preferences and features

alimirjamali committed:
Improve qvm-ls options to avoid parsing

fixes: https://github.com/QubesOS/qubes-issues/issues/8614

Filtering, Formatting & Sorting ArgumentParser Groups
Sorting options based on output columns
A new pref(erence) format
Dedicated filter options for class, label, template, netvm preferences
Dedicated filter options for internal,servicevm & updates-available
Generic filtering options for preferences and features

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -330,7 +330,10 @@ def calc_used(vm, volume_name):
-# todo maxmem
+# Show hyphen if maxmmem is zero (for HVMs)
+Column('MAXMEM',
+    attr=(lambda vm: vm.maxmem if vm.virt_mode != "hvm" else '-'),
```
1. Why use a symbol `-` rather than `0` itself?
2. It doesn't _Show hyphen if maxmmem is zero (for HVMs)_, but it does show hyphen if `virt_mode` is `hvm`.

alimirjamali replied:
> 1. Why use a symbol `-` rather than `0` itself?

I also considered 'N/A'. I personally thought hyphen or N/A should imply as not-applicable as maxmem does not make sense for HVMs. There are similar cases for the other fields.

marmarek replied:
maxmem very much does make sense for HVM. It doesn't make sense for qubes with dynamic memory management disabled, which happen to be the case for the only HVMs in the default configuration.

alimirjamali replied:
I think it is better to remove maxmem custom implementation from this PR since it is not relevant to the original target of the PR.

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -496,6 +511,8 @@ def write_table(self, stream=sys.stdout):
     'kernel': ('name', 'state', 'class', 'template', 'kernel', 'kernelopts'),
     'full': ('name', 'state', 'class', 'label', 'qid', 'xid', 'uuid'),
 #   'perf': ('name', 'state', 'cpu', 'memory'),
```
Makes sense to delete this comment as you added a substitute in the line below.

alimirjamali replied:
> Makes sense to delete this comment as you added a substitute in the line below.

They are different. One is perf(ormance). The other is pref(erence). A little bit confusion I guess. Maybe one should be made longer.

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -594,60 +611,115 @@ def get_parser():
-    parser.add_argument('--help-columns', action=_HelpColumnsAction)
-    parser.add_argument('--help-formats', action=_HelpFormatsAction)
+    parser_format = parser.add_argument_group(title='formatting options')
+    parser_format_group = parser_format.add_mutually_exclusive_group()
```
This name is better than `parser_formats`. Maybe use `parser_exclusive_format`. Nitpick.

alimirjamali replied:
> This name is better than `parser_formats`. Maybe use `parser_exclusive_format`. Nitpick.

Makes very sense. I will apply it after receiving other feedback from others (maybe Marek).

alimirjamali commented:
Making pylint happy(er)

I am waiting for the CI/CD bot to advise on the required unit tests.
The Sphinx error could be resolved by updating documentation if the main change is approved. 
```
18:14:05,477 [executor:docker:b34f76404a9f] output: DEBUG: Sphinx error:
18:14:05,478 [executor:docker:b34f76404a9f] output: DEBUG: Undocumented arguments for command 'qvm-ls': '--based-on, --class, --conn-netvm, --features, --ignore-case, --internal, --label, --no-tags, --pending-update, --prefs, --reverse, --servicevm, --sort'
```

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -594,60 +611,115 @@ def get_parser():
+
+    parser_filter.add_argument('--based-on', nargs='+',
+        metavar='TEMPLATE', action='store',
+        help='filter results to the AppVMs based on the TEMPLATE. '
```
DispVMs are not AppVMs have an AppVM as template, therefore don't use _AppVMs based on_, but use _qubes based on_.

alimirjamali replied:
Theoretically DispVMs are not directly based on a TemplateVM. But rather based on an AppVM which is transformed to be a DispVMTemplate. But Yes. You are right. I will change it _qubes based on_.

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -594,60 +611,115 @@ def get_parser():
+        parser_filter.add_argument('--{}'.format(pwstate), action='store_true',
+        help='show {} VMs'.format(pwstate))
+
+    parser_filter.add_argument('--based-on', nargs='+',
```
Name should use qubes nomenclature, `--template-source` maybe? Unsure, possibly a nitpick in this case, will let others step in.

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -594,60 +611,115 @@ def get_parser():
+
+    parser_filter.add_argument('--conn-netvm', nargs='+',
+        metavar='NETVM', action='store',
+        help='filter results to the VMs connecting via NETVM')
```
Prefer. `qubes` instead of `VMs`.

From the [Qubes OS glossary](https://www.qubes-os.org/doc/glossary/#qube):

> A secure compartment in Qubes OS. Currently, qubes are implemented as Xen VMs, but Qubes OS is independent of its underlying compartmentalization technology. VMs could be replaced with a different technology, and qubes would still be called “qubes.”

I notice that there are existing csae of `VM` word being used, but please don't follow it.

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -594,60 +611,115 @@ def get_parser():
+        help='filter results to the AppVMs based on the TEMPLATE. '
+             '"" means None')
+
+    parser_filter.add_argument('--conn-netvm', nargs='+',
```
Please avoid abbreviations unless widely know. I know `conn` stands for connection, but not everybody may know.

It is from [Qubes OS website style guide](https://www.qubes-os.org/doc/website-style-guide/#general-practices), but I believe it applies to documentation also.

> Use descriptive variable names. Never use one or two letter variable names. Avoid obscure abbreviations and made-up words.

A better name would be `--netvm-is` to not conflict with a possible future option `--netvm`. Not that my name `--netvm-is` is great... but out of ideas now.

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -594,60 +611,115 @@ def get_parser():
+    parser_filter.add_argument('--internal', metavar='<y|n|both>',
+        default='both', action='store', choices=['y', 'yes', 'n', 'no', 'both'],
+        help='show only internal VMs or option to hide them. '
+             'default is showing both regular & internal VMs')
```
If default is showing both, why have a `both` option?

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -594,60 +611,115 @@ def get_parser():
+             'default is showing both regular & internal VMs')
+
+    parser_filter.add_argument('--servicevm', metavar='<y|n|both>',
+        default='both', action='store', choices=['y', 'yes', 'n', 'no', 'both'],
```
Same for `both` here.

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -594,60 +611,115 @@ def get_parser():
+
+    parser_filter.add_argument('--prefs', nargs='+', metavar='PREFERENCE=VALUE',
+        action='store',
+        help='filter results to VMs with all of specified preferences. '
```
_to qubes that matches all specified_

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +782,109 @@ def main(args=None, app=None):
+        spinner.show('Filtering based on labels...')
+        for dom in domains:
+            if dom.label.name in args.label:
+                domains_labeled.append(dom)
```
Why outside of list comprehension like in `args.klass`? Was this for the spinner updates?

alimirjamali replied:
> Was this for the spinner updates?

Yes. Exactly. If qvm-ls is run outside of dom0, the policy checks comes into play and it could become slow. And it could make it slow. I  considered moving it to Table class. But there might be cases that a field is required for filtering but not in output.

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +782,109 @@ def main(args=None, app=None):
+        spinner.show('Filtering results to VMs based on their template...')
+        for dom in domains:
+            if getattr(dom, 'template', '') in args.based_on:
+                child_domains.append(dom)
```
Why outside of list comprehension like in args.klass? Was this for the spinner updates?

alimirjamali replied:
Same as  above

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +782,109 @@ def main(args=None, app=None):
+        spinner.show('Filtering results to VMs based on their netvm...')
+        for dom in domains:
+            if getattr(dom, 'netvm', '') in args.conn_netvm:
+                domains_connecting.append(dom)
```
Why outside of list comprehension like in args.klass? Was this for the spinner updates?

alimirjamali replied:
same as above

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +782,109 @@ def main(args=None, app=None):
+
+    if args.pending_update:
+        domains = [d for d in domains if
+                   d.features.get('updates-available', None)]
```
That is very nice, simple and nice to have.

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +786,109 @@ def main(args=None, app=None):
+            try:
+                key, value = pref.split('=', 1)
+            except ValueError:
+                parser.error("Invalid argument --prefs {}".format(pref))
```
Add a colon to separate the message from the argument: `Invalid argument: --prefs {}`

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +786,109 @@ def main(args=None, app=None):
+            except ValueError:
+                parser.error("Invalid argument --prefs {}".format(pref))
+            if not key:
+                parser.error("Invalid argument --prefs {}".format(pref))
```
Add a colon to separate the message from the argument: `Invalid argument: --prefs {}`

alimirjamali commented:
Making pylint even happier (complain on assignment of lambda to variable)

The CI/CD build failure logs are very cluttered and hard to understand. If it is because of sphinx complain, should be fixable by updating the documents (if the overall changes and new options are acceptable).

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -479,6 +486,18 @@ def write_table(self, stream=sys.stdout):
                     except qubesadmin.exc.QubesVMNotFoundError:
                         continue
+                if self.sort_order in self.get_head():
+                    sort_index = self.get_head().index(self.sort_order)
```
Store `get_head()` output to a variable to avoid repeated calls that will always have the same value.

alimirjamali replied:
It is only called again once (and not in a loop). I guess we could pass this

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -479,6 +486,18 @@ def write_table(self, stream=sys.stdout):
+                        table_data[1:] = \
+                            sorted(table_data[1:],
+                                   key=(lambda row: row[sort_index]),
+                                   reverse=self.reverse_sort)
```
Check for `ignore_case` inside the lambda, is this more readable?
```python3
table_data[1:] = sorted(table_data[1:],
    key=(lambda row: row[sort_index].upper()) if self.ignore_case else row[sort_index],
    reverse=self.reverse_sort)
```

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -594,60 +615,115 @@ def get_parser():
+    parser_filter.add_argument('--tags', nargs='+', metavar='TAG',
+        help='show only VMs having specific tag(s)')
+
+    parser_filter.add_argument('--no-tags', nargs='+', metavar='TAG',
```
I understand many unix tools has the negating option prefix `--no-something`, but it doesn't mean it will print qubes that have no tags, but qubes that don't have a certain tag. For this reason and because `qvm_ls` already has the `--exclude` option, please change `--no-tags` to `--exclude-tags`.

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +786,109 @@ def main(args=None, app=None):
+
+    if args.conn_netvm:
+        # Filter only VMs connecting with specific netvm
+        domains_connecting = []
```
The `domains_connecting` is not a good variable name, there can be multiple types of qube connections such as audiovm, netvm, guivm. Please use a name that has the connection name such as `domains_child_of_netvm` or something similar that has `netvm` in the name.

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +786,109 @@ def main(args=None, app=None):
+                value = None
+            elif value in ['\'\'', '""']:
+                value = ''
+            domains = [d for d in domains if str(getattr(d, key, None))==value]
```
Missing space around test: `) == value`.

alimirjamali replied:
> Missing space around test: `) == value`.

I tried to cheat here and avoid breaking the line above 80 characters. But yes, I should do it.

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +786,109 @@ def main(args=None, app=None):
+                value = None
+            elif value in ['\'\'', '""']:
+                value = ''
+            domains = [d for d in domains if d.features.get(key, None) == value]
```
Have you benchmarked this? Same thing about `args.prefs` later on. It may be slow, but decreasingly slow as the `domains` supposedly gets smaller.

Validating the arguments in the for loop seems okay, but the features.get() is ran for every feature instead of once per feature key pair.

If the runtime is not bad, then may be okay to leave as it is.

alimirjamali replied:
> Have you benchmarked this?

They are almost instant in dom0 on this HP Elitebook 820 G1 from Windows 7 era. But noticeable in other domains (as policy checks come into play). Not painful. Very much usable. Maybe I should consider adding the spinner for features and preferences filtering as well.

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +786,109 @@ def main(args=None, app=None):
+
+    if args.based_on:
+        # Filter only VMs based on specific TemplateVM
+        child_domains = []
```
Please use a name that reference the it is child domain of a template. The function `tree_append_child()` uses `childs` variable for network children. Use something similar to `domains_child_of_template`. Maybe child is redundant in this name.

Also you use `child_domains` here and `domains_connecting` later. Use a consistent naming scheme, such as beginning always with `domains_` and referencing the connection type later.

ben-grande reviewed (changes requested):
Good work!

alimirjamali commented:
TODo: Numeric sort wherever applicable. ~~Will be implemented after receiving further feedback from others.~~ Implemented.

alimirjamali committed:
Add requested changes for qvm-ls, numeric sort

Requested changes by ben-grande (#301)
Numeric sort if applicable

alimirjamali committed:
Update documentation for new qvm-ls options

Follow-up to #301
Options are grouped as general, formatting, filtering and sorting

alimirjamali commented:
Ok. After updating the documentation, we are past CI/CD bot (all Green) and Codecov is advising on the required unit tests. I will write them next.

alimirjamali committed:
Update unittests for new qvm-ls options

Follow-up to #301
Plus minor fix for documentation

ben-grande commented on doc/manpages/qvm-ls.rst:
```diff
@@ -79,21 +102,83 @@ Options
+
+.. option:: --tags TAG ...
+
+   Shows only VMs having specific tag(s).
```
`s/VMs/qubes/`

ben-grande commented on doc/manpages/qvm-ls.rst:
```diff
@@ -79,21 +102,83 @@ Options
+
+.. option:: --exclude-tags TAG ...
+
+   Exclude VMs having specific tag(s).
```
`s/VMs/qubes/`

ben-grande commented on doc/manpages/qvm-ls.rst:
```diff
@@ -79,21 +102,83 @@ Options
+
+.. option:: --running, --paused, --halted
+
+   Shows only VMs matching the specified power state(s). When none of these
```
`s/VMs/qubes/`

ben-grande commented on doc/manpages/qvm-ls.rst:
```diff
@@ -79,21 +102,83 @@ Options
+.. option:: --running, --paused, --halted
+
+   Shows only VMs matching the specified power state(s). When none of these
+   options is used (default), all VMs are shown.
```
`s/VMs/qubes/`

ben-grande commented on qubesadmin/tests/tools/qvm_ls.py:
```diff
@@ -355,3 +363,347 @@ def test_101_list_selected(self):
+                ('a', TestVM('a', label='red', maxmem='100')),
+                ('B', TestVM('B', label='green', maxmem='1000')),
+                ('c', TestVM('c', label='blue',  maxmem='300')),
+                ('dom0', TestVM('dom0', label='black', maxmem='-'))
```
Is this maxmem value `-` legacy from your previous implementation of returning `-` if it is `0`? I still prefer integer value.

alimirjamali replied:
No. It is the current existing condition. Not related to my previous (or current) implementation. This is only the unit-test BTW. I had to assure that the numeric sort works the `-`

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -594,60 +626,113 @@ def get_parser():
+
+    parser_filter.add_argument('--servicevm', metavar='<y|n>',
+        default='both', action='store', choices=['y', 'yes', 'n', 'no'],
+        help='show only Service VMs or option to hide them')
```
`s/Service VMs/ServiceVMs/`

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -594,60 +626,113 @@ def get_parser():
+
+    parser_filter.add_argument('--netvm-is', nargs='+',
+        metavar='NETVM', action='store',
+        help='filter results to the qubes connecting via NETVM')
```
The template filter above has:

> filter results to the qubes based on the TEMPLATE.

So for uniformity, I propose:

> filter results to the qubes based on the NETVM.

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -594,60 +626,113 @@ def get_parser():
+        help='reverse sort')
+
+    parser_sort.add_argument('--ignore-case', action='store_true',
+        default=False, help='ignore case distinctions for sorting')
```
`s/for sorting/when sorting/`

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +795,110 @@ def main(args=None, app=None):
         ]
 
+    if args.klass:
+        # filter only VMs to specific class(es)
```
`s/VMs/qubes/`

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +795,110 @@ def main(args=None, app=None):
+        domains = [d for d in domains if d.klass in args.klass]
+
+    if args.label:
+        # filter only VMs with specific label(s)
```
`s/VMs/qubes/`

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +795,110 @@ def main(args=None, app=None):
+        spinner.hide()
+
     if args.tags:
         # filter only VMs having at least one of the specified tags
```
`s/VMs/qubes/`

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +795,110 @@ def main(args=None, app=None):
                    if set(dom.tags).intersection(set(args.tags))]
 
+    if args.exclude_tags:
+        # exclude VMs having at least one of the specified tags
```
`s/VMs/qubes/`

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +795,110 @@ def main(args=None, app=None):
+                   if not set(dom.tags).intersection(set(args.exclude_tags))]
+
+    if args.template_source:
+        # Filter only VMs based on specific TemplateVM
```
`s/VMs/qubes/`

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +795,110 @@ def main(args=None, app=None):
+        spinner.hide()
+
+    if args.netvm_is:
+        # Filter only VMs connecting with specific netvm
```
`s/VMs/qubes/`

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +795,110 @@ def main(args=None, app=None):
+                   d.features.get('updates-available', None)]
+
+    if args.features:
+        # Filter only VMs with specified features
```
`s/VMs/qubes/`

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -710,16 +795,110 @@ def main(args=None, app=None):
+            domains = [d for d in domains if d.features.get(key, None) == value]
+
+    if args.prefs:
+        # Filter only VMs with specified preferences
```
`s/VMs/qubes/`

ben-grande commented on doc/manpages/qvm-ls.rst:
```diff
@@ -79,21 +102,83 @@ Options
+.. option:: --features FEATURE=VALUE ...
+
+   Filter results to qubes that match all specified features. Omitted VALUE
+   means None (not set). "" or '' means blank
```
> Omitted VALUE means None (not set). "" or '' means blank

What about:

_Omitted VALUE means None (unset). Empty value means "" or '' (blank)_

ben-grande commented on doc/manpages/qvm-ls.rst:
```diff
@@ -79,21 +102,83 @@ Options
+.. option:: --prefs PREFERENCE=VALUE ...
+
+   Filter results to qubes that match all specified preferences. Omitted VALUE
+   means None (not set). "" or '' means blank
```
> Omitted VALUE means None (not set). "" or '' means blank

What about:

_Omitted VALUE means None (unset). Empty value means "" or '' (blank)_

ben-grande commented on doc/manpages/qvm-ls.rst:
```diff
@@ -79,21 +102,83 @@ Options
+
+.. option:: --sort COLUMN
+
+   Sort based on provided column rather than NAME. Sort key should be in the
```
`s/should/must/` if it is required.

ben-grande commented on doc/manpages/qvm-ls.rst:
```diff
@@ -79,21 +102,83 @@ Options
+
+.. option:: --ignore-case
+
+   Ignore case distinctions for sorting
```
`s/for sorting/when sorting/`

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -594,60 +626,113 @@ def get_parser():
+    parser_filter.add_argument('--template-source', nargs='+',
+        metavar='TEMPLATE', action='store',
+        help='filter results to the qubes based on the TEMPLATE. '
+             '"" means None')
```
> Omitted VALUE means None (not set). "" or '' means blank

What about:

_Omitted VALUE means None (unset). Empty value means "" or '' (blank)_

alimirjamali replied:
This one is tricky. A template name could not be blank. This is a case that we could have only None.

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -594,60 +626,113 @@ def get_parser():
+    parser_filter.add_argument('--features', nargs='+', metavar='FEATURE=VALUE',
+        action='store',
+        help='filter results to qubes that matches all specified features. '
+            'omitted VALUE means None. "" means blank')
```
> Omitted VALUE means None (not set). "" or '' means blank

What about:

_Omitted VALUE means None (unset). Empty value means "" or '' (blank)_

ben-grande commented on qubesadmin/tools/qvm_ls.py:
```diff
@@ -594,60 +626,113 @@ def get_parser():
+    parser_filter.add_argument('--prefs', nargs='+', metavar='PREFERENCE=VALUE',
+        action='store',
+        help='filter results to qubes that matches all specified preferences. '
+            'omitted VALUE means None. "" means blank')
```
> Omitted VALUE means None (not set). "" or '' means blank

What about:

_Omitted VALUE means None (unset). Empty value means "" or '' (blank)_

ben-grande commented on doc/manpages/qvm-ls.rst:
```diff
@@ -43,14 +75,10 @@ Options
 
-   Shows only VMs matching the specified power state(s). When none of these
-   options is used (default), all VMs are shown.
+   List domains as a network tree. Domains are sorted as they are connected to
```
`s/domains/qubes/`, `s/Domains/qubes/`

ben-grande reviewed (changes requested):
Documentation review.

marmarek reviewed (approved):
Looks okay, but just in case lets run it through openqa.

alimirjamali committed:
Improve new qvm-ls documentation & comments

alimirjamali commented:
A supplementary commit to implement documentation & commenting improvements based on @ben-grande suggestions
